src/libstd/num/f64.rs

```diff
@@ -546,7 +567,27 @@ impl f64 {
     #[stable(feature = "rust1", since = "1.0.0")]
     #[inline]
     pub fn log2(self) -> f64 {
-        unsafe { intrinsics::log2f64(self) }
+        if !cfg!(target_os = "sunos") {
+            unsafe { intrinsics::log2f64(self) }
+        } else {
+            // Illumos requires a wrapper around the log2 function because of
+            // its non-standard behavior
+            if self.is_finite() {
+                if self > 0.0 {
+                    unsafe { intrinsics::log2f64(self) }
+                } else if self == 0.0 {
+                    NEG_INFINITY // log2(0) = -Inf
+                } else {
+                    NAN // log2(-n) = NaN
+                }
+            } else if self.is_nan() {
+                self // log2(NaN) = NaN
+            } else if self > 0.0 {
+                self // log2(Inf) = Inf
+            } else {
+                NAN // log2(-Inf) = NaN
+            }
```

Comment:
It hasn't worked with a generic signature `fn foo<F: Fn(f64) -> f64>(val: f64, f: F) -> f64`, because the intrinsic functions are unsafe & extern - they don't implement the Fn trait. However, it worked this way: `fn foo(self, log_fn: unsafe extern "rust-intrinsic" fn(f64) -> f64) -> f64` - i.e., with a completely defined type in a function reference.

Would this work as a solution, or should I find a better way?